src/librustc_mir/transform/const_prop.rs

```diff
@@ -55,14 +54,11 @@ impl<'tcx> MirPass<'tcx> for ConstProp {
             .expect("Non-local call to local provider is_const_fn");
 
         let is_fn_like = FnLikeNode::from_node(tcx.hir().get(hir_id)).is_some();
-        let is_assoc_const = match tcx.def_kind(source.def_id()) {
-            Some(DefKind::AssocConst) => true,
-            _ => false,
-        };
 
-        // Only run const prop on functions, methods, closures and associated constants
-        if !is_fn_like && !is_assoc_const {
-            // skip anon_const/statics/consts because they'll be evaluated by miri anyway
+        // Only run const prop on functions, methods, closures.
+        if !is_fn_like {
+            // Skip anonymous consts, statics and consts because they'll be evaluated by miri
+            // anyway; skip associated consts to avoid duplicating errors.
             trace!("ConstProp skipped for {:?}", source.def_id());
             return;
         }
```

Comment:
I've done this to avoid emitting overflow diagnostics twice (see https://github.com/rust-lang/rust/pull/70017/commits/908b5e1176931692ca637d693ff152e0175a0b60, which shows how the stderrs are without this change). It's evident that const_prop was deliberately being run for associated consts, but I've not noticed any changes to the generated mir in some small test cases. Are there any glaring issues with this change? Is there a particular reason why associated consts don't also fall into the category of being evaluated by miri anyway?

Reply:
associated consts can depend on generic parameters, thus you may not be able to actually evaluate them.

Reply:
The reason you get duplicate errors is that they get evaluated by Miri anyway. So, it makes no sense to separate "anonymous consts, statics and consts" from "assoc consts" in this list.

However, as @oli-obk said, this will regress diagnostics for assoc consts that cannot be evaluated. If you adjust the `issue-69020` test to commute the operands around (`-i32::MIN + T::NEG` -> `T::NEG + (-i32::MIN)`), do you even still get the errors?

Reply:
Yeah, in this case, I still get the errors even after commuting the operands around.

I can't think of any particular cases where diagnostics will regress by this change, but I think I see your point. It's not unreasonable to me that there must exist some case where `const_eval_poly` fails to evaluate anything without any substitutions that const eval _with_ substitutions would succeed.

Skipping const_prop on anon consts just to avoid duplicating errors doesn't seem like an elegant solution anyway. I only did this because I noticed this is how consts didn't duplicate on errors -- but then again, unlike assoc consts, normal consts don't ever need to handle generics. Yeah -- I think I'll pursue https://github.com/rust-lang/rust/pull/70017#discussion_r392726654 and go the opposite approach - keep the error emitted by run of const_prop, and don't emit the duplicate error emitted by the run of check_const in UnusedBrokenConst.

Reply:
> Yeah, in this case, I still get the errors even after commuting the operands around.

Can you explain why that is? Const-eval cannot run on those consts as the generic value is not known, and you disabled const-prop for them.

I am pretty sure this change here will regress diagnostics by not running on assoc consts, we just need to find the right test case for it.

Reply:
Yeah, const_prop doesn't run on associated consts, but with my changes, the UnusedBrokenConst lint does. Const eval runs because UnusedBrokenConst calls `const_eval_poly` to do its const checking. `(-i32::MIN)` can still be linted against because it doesn't depend on any generic parameters (and, from what I can see, negation of the MIN value is specifically checked for). Same thing for the other values: we can still lint because we know for any `T:  Foo`, no matter the actual type of `T`, `<T as Foo>::{NEG, ADD, DIV, OOB}` is an `i32` -- const eval isn't really dependent on the actual value of `T` for this case.

Reply:
But if the code is `T::NEG + ...`, then it should *first* evaluate `T::NEG` before it even gets to evaluate the overflowing arithmetic. And that evaluation can only fail as `T` is not know.

So, I don't see how const-eval can even detect the overflow here. It proceed statement by statement through the const initializer, and the statement that evaluates `T::NEG` should come before the one that overflows -- so the latter is never reached.

Reply:
Ah, I see now. `-i32::MIN` is evaluated first, no matter which way I put the operands. There must be some cleverness going on here, because it appears first in the generated mir no matter which way I put them.

 If I do the following, though,
```rust
use std::i32;
pub trait Foo {
    const NEG: i32;
}

impl<T: Foo> Foo for Vec<T> {
    const NEG: i32 = --T::NEG + (-i32::MIN);
}
```
I manage to get `T::NEG` placed first in the mir, and then yeah -- I don't emit a diagnostic for `-i32::MIN` anymore :slightly_frowning_face:. I'll go ahead and drop this change and fix the duplicate errors properly, then.

Reply:
Nice test case reduction. :)  Please also add the changed test to the test suite.